### PR TITLE
[flutter_tools] do not error doctor on missing vs code extension

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

In cases where the VSCode plugins are not located, display links to where they can be downloaded but do not surface an error.

```
PS C:\Users\Jonah\flutter> flutter doctor -v
[√] Flutter (Channel master, 1.22.0-10.0.pre.372, on Microsoft Windows [Version 10.0.19041.508], locale en-US)
    • Flutter version 1.22.0-10.0.pre.372 at C:\Users\Jonah\flutter
    • Framework revision 99d0954857 (2 days ago), 2020-09-25 14:29:22 -0700
    • Engine revision 3a73d073c8
    • Dart version 2.11.0 (build 2.11.0-161.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at C:\Users\Jonah\AppData\Local\Android\sdk
    • Platform android-29, build-tools 29.0.3
    • ANDROID_HOME = C:\Users\Jonah\AppData\Local\Android\sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.7.30517.126
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (version 3.6)
    • Android Studio at C:\Program Files\Android\Android Studio
    X Flutter plugin not installed; this adds Flutter specific functionality.
    X Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)

[√] VS Code (version 1.49.2)
    • VS Code at C:\Users\Jonah\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension can be installed from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[√] Connected device (4 available)
    • Windows (desktop) • windows    • windows-x64    • Microsoft Windows [Version 10.0.19041.508]
    • Web Server (web)  • web-server • web-javascript • Flutter Tools
    • Chrome (web)      • chrome     • web-javascript • Google Chrome 85.0.4183.121
    • Edge (web)        • edge       • web-javascript • Microsoft Edge 85.0.564.51
```

### CODE REVIEW

**Comment by DanTup** [comment]
This was changed in https://github.com/flutter/flutter/pull/25269 to be consistent with Android Studio (reported at https://github.com/flutter/flutter/issues/22931). Should we keep them consistent? (I don't have a strong preference, though since it was raised before I figured it worth asking).

**Comment by jonahwilliams** [comment]
I'm also changing AS and intellij

**Review by DanTup — APPROVED** [review]

**Comment by DanTup** [comment]
> I'm also changing AS and intellij

Ah, sgtm! (and lgtm!)

File: packages/flutter_tools/lib/src/vscode/vscode.dart [hunk 1]
```diff
@@ -49,40 +49,39 @@ class VsCode {
     if (extensionDirs.isNotEmpty) {
       final FileSystemEntity extensionDir = extensionDirs.first;
 
-      _isValid = true;
       _extensionVersion = Version.parse(
           extensionDir.basename.substring('$extensionIdentifier-'.length));
       _validationMessages.add(ValidationMessage('Flutter extension version $_extensionVersion'));
     } else {
       _validationMessages.add(notInstalledMessage);
+      return;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/vscode/vscode.dart:57]: nit: return not needed.

File: packages/flutter_tools/lib/src/vscode/vscode.dart [hunk 2]
```diff
@@ -31,16 +31,16 @@ class VsCode {
 
     // If the extensions directory doesn't exist at all, the listSync()
     // below will fail, so just bail out early.
-    const ValidationMessage notInstalledMessage = ValidationMessage.error(
-          'Flutter extension not installed; install from\n$extensionMarketplaceUrl');
-    if (!globals.fs.isDirectorySync(extensionDirectory)) {
+    const ValidationMessage notInstalledMessage = ValidationMessage(
+          'Flutter extension can be installed from\n$extensionMarketplaceUrl');
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/vscode/vscode.dart:35]: nit: two space indent on continued line.

File: packages/flutter_tools/lib/src/vscode/vscode.dart [hunk 3]
```diff
@@ -259,16 +260,15 @@ class VsCode {
...
+    if (!fileSystem.isFileSync(packageJsonPath)) {
       return null;
     }
-    final String jsonString = globals.fs.file(packageJsonPath).readAsStringSync();
+    final String jsonString = fileSystem.file(packageJsonPath).readAsStringSync();
     try {
       final Map<String, dynamic> jsonObject = castStringKeyedMap(json.decode(jsonString));
       return jsonObject['version'] as String;
-    } on FormatException catch (err) {
-      globals.printTrace('Error parsing VSCode $packageJsonPath:\n$err');
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/vscode/vscode.dart:271]: Could this still be useful?

**Review by zanderso — APPROVED** [review]
lgtm w/ nits

File: packages/flutter_tools/lib/src/vscode/vscode.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vscode/vscode.dart:271]: I was only using this as a sentinel in a test, but not throwing an exception accomplishes the same goal

File: packages/flutter_tools/lib/src/vscode/vscode.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vscode/vscode.dart]: Done

File: packages/flutter_tools/lib/src/vscode/vscode.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/vscode/vscode.dart]: Done

**Comment by jonahwilliams** [comment]
G3fix seems to be a bit confused on this one. the latest failure is due to it missing a CL already merged

**Pull request merged**